cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/services/ApplicationLifecycleService.java

```diff
@@ -474,6 +475,37 @@ public void upgrade(boolean continueOnFailure) throws Exception {
     }
   }
 
+  // look up the app archive location from the store.  Most of this logic is in case that jar isn't actually
+  // there. In that case we try to find it in the expected place.
+  private Location findAppJarLocation(Id.Application appId) throws IOException {
+    Location appJarLocation = store.getApplicationArchiveLocation(appId);
+    if (appJarLocation == null) {
+      return null;
+    }
+    // bad metadata... not sure how it gets into this state but we have seen it
+    if (!appJarLocation.exists()) {
+      // make an educated guess for where it could be
+      appJarLocation =
+        ProgramGenerationStage.getAppArchiveDirLocation(configuration, appId, namespacedLocationFactory);
+      if (!appJarLocation.exists() || !appJarLocation.isDirectory()) {
+        return null;
+      }
+      // should be only one file there... expect it to start with the app name and end in .jar
+      boolean found = false;
+      for (Location file : appJarLocation.list()) {
+        if (file.getName().startsWith(appId.getId()) && file.getName().endsWith(".jar")) {
+          appJarLocation = file;
+          found = true;
+          break;
+        }
+      }
+      if (!found) {
+        return null;
+      }
+    }
+    return appJarLocation;
```

Comment:
It seems that instead of returning null, this method should throw an exception with a detailed message, why the location could not be found (there can be various reasons). The caller can catch the exception and log it properly.